cranelift/codegen/src/isa/zkasm/inst/emit.rs

```diff
@@ -1632,24 +1632,33 @@ impl MachInstEmit for Inst {
                 debug_assert_eq!(a, a0());
                 debug_assert_eq!(b, b0());
 
-                let opcode = match cc {
-                    IntCC::Equal => "EQ",
-                    IntCC::NotEqual => "EQ",
-                    IntCC::SignedLessThan => "SLT",
-                    IntCC::SignedGreaterThanOrEqual => todo!(),
-                    IntCC::SignedGreaterThan => todo!(),
-                    IntCC::SignedLessThanOrEqual => todo!(),
-                    IntCC::UnsignedLessThan => "LT",
-                    IntCC::UnsignedGreaterThanOrEqual => todo!(),
-                    IntCC::UnsignedGreaterThan => todo!(),
-                    IntCC::UnsignedLessThanOrEqual => todo!(),
+                let (opcode, invert_result, swap_inputs) = match cc {
+                    IntCC::Equal => ("EQ", false, false),
+                    IntCC::NotEqual => ("EQ", true, false),
+                    IntCC::SignedLessThan => ("SLT", false, false),
+                    IntCC::SignedGreaterThanOrEqual => ("SLT", true, false),
+                    IntCC::SignedGreaterThan => ("SLT", false, true),
+                    IntCC::SignedLessThanOrEqual => ("SLT", true, true),
+                    IntCC::UnsignedLessThan => ("LT", false, false),
+                    IntCC::UnsignedGreaterThanOrEqual => ("LT", true, false),
+                    IntCC::UnsignedGreaterThan => ("LT", false, true),
+                    IntCC::UnsignedLessThanOrEqual => ("LT", true, true),
                 };
 
-                put_string(&format!("$ => {} :{opcode}\n", reg_name(rd.to_reg())), sink);
+                // NB: We can implement this more efficiently by either introducing a dedicated
+                // ZK ASM instruction or introducing a new ISLE instruction and asking register
+                // allocator to put the inputs in the reverse from the start.
+                if swap_inputs {
+                    put_string(&format!("A => C\n"), sink);
+                    put_string(&format!("B => A\n"), sink);
+                    put_string(&format!("C => B\n"), sink);
```

Comment:
A bit surprised that zkasm don't allow ``A, B => B, A``, but tried, it don't work unfortunately. But ``C => A, B`` works, and takes 2 cycles less (which is a lot). Is it possible to implement such swaps for zkasm?..

Reply:
Yeah, I've tried a few things but haven't found a magic incantation for this. Let's ask Polygon folks on Monday.